File: Module - RESTful Web Services/Rest/restserver/src/main/java/com/javadude/rest/server/TodoItem.java
package com.javadude.rest.server;

import org.json.JSONObject;

public class TodoItem {
	private long id;
	private String name;
	private String description;
	private int priority;

	public long getId() {
		return id;
	}

	public void setId(long id) {
		this.id = id;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getDescription() {
		return description;
	}

	public void setDescription(String description) {
		this.description = description;
	}

	public int getPriority() {
		return priority;
	}

	public void setPriority(int priority) {
		this.priority = priority;
	}

	public TodoItem(long id, String name, String description, int priority) {
		this.id = id;
		this.name = name;
		this.description = description;
		this.priority = priority;
	}

	@Override
	public String toString() {
		return "TodoItem{" +
				"id=" + id +
				", name='" + name + '\'' +
				", description='" + description + '\'' +
				", priority=" + priority +
				'}';
	}

	public String toJsonString() {
		JSONObject jsonObject = new JSONObject();
		jsonObject.put("id", id);
		jsonObject.put("name", name);
		jsonObject.put("description", description);
		jsonObject.put("priority", priority);
		return jsonObject.toString(4);
	}
}

File: Module - RESTful Web Services/Rest2/data/src/main/java/com/javadude/data/TodoItem.kt
package com.javadude.data

import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.builtins.list
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonConfiguration
import java.util.UUID

@Serializable
data class TodoItem(
    var id : String = UUID.randomUUID().toString(),
    var name : String = "",
    var description : String = "",
    var priority : Int = 1
)

private val json = Json { allowStructuredMapKeys = true }

// NOTE - the Kotlin Serialization lib has changed the API since the video was created
//    the code below correctly works with the 1.0 RC version of kotlin serialization
fun TodoItem.toJsonString() =
    json.encodeToString(TodoItem.serializer(), this)

fun List<TodoItem>.todoListToJsonString() =
    json.encodeToString(ListSerializer(TodoItem.serializer()), this)

fun String.toTodoItem() =
    if (this.isEmpty()) null else json.decodeFromString(TodoItem.serializer(), this)

fun String.toTodoItemList() =
    if (this.isEmpty()) emptyList() else json.decodeFromString(ListSerializer(TodoItem.serializer()), this)

fun ByteArray.toTodoItem() = String(this).toTodoItem()
fun ByteArray.toTodoItemList() = String(this).toTodoItemList()

